code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm

```diff
@@ -0,0 +1,367 @@
+	var/interactive = TRUE // So mapmakers can disable interaction.
+	var/cooling = TRUE
+	var/base_heating = 140
+	var/base_cooling = 170
+	var/obj/item/tank/holding
+
+/obj/machinery/atmospherics/components/binary/thermomachine/Initialize()
+	. = ..()
+	RefreshParts()
+	update_appearance()
+
+/obj/machinery/atmospherics/components/binary/thermomachine/getNodeConnects()
+	return list(dir, turn(dir, 180))
+
+/obj/machinery/atmospherics/components/binary/thermomachine/proc/swap_function()
+	cooling = !cooling
+	if(cooling)
+		icon_state_off = "freezer"
+		icon_state_on = "freezer_1"
+		icon_state_open = "freezer-o"
+	else
+		icon_state_off = "heater"
+		icon_state_on = "heater_1"
+		icon_state_open = "heater-o"
+	target_temperature = T20C
+	RefreshParts()
+	update_appearance()
+
+/obj/machinery/atmospherics/components/binary/thermomachine/on_construction(obj_color, set_layer)
+	var/obj/item/circuitboard/machine/thermomachine/board = circuit
+	if(board)
+		piping_layer = board.pipe_layer
+		set_layer = piping_layer
+
+	for(var/obj/machinery/atmospherics/device in get_turf(src))
+		if(device.piping_layer != piping_layer || device == src)
+			continue
+		visible_message("<span class='warning'>A pipe is hogging the output, remove the obstruction or change the machine piping layer.</span>")
+		deconstruct(TRUE)
+		return
+	return..()
+
+/obj/machinery/atmospherics/components/binary/thermomachine/RefreshParts()
+	var/calculated_bin_rating
+	for(var/obj/item/stock_parts/matter_bin/bin in component_parts)
+		calculated_bin_rating += bin.rating
+	heat_capacity = 5000 * ((calculated_bin_rating - 1) ** 2)
+	min_temperature = T20C
+	max_temperature = T20C
+	if(cooling)
+		var/calculated_laser_rating
+		for(var/obj/item/stock_parts/micro_laser/laser in component_parts)
+			calculated_laser_rating += laser.rating
+		min_temperature = max(T0C - (base_cooling + calculated_laser_rating * 15), TCMB) //73.15K with T1 stock parts
+	else
+		var/calculated_laser_rating
+		for(var/obj/item/stock_parts/micro_laser/laser in component_parts)
+			calculated_laser_rating += laser.rating
+		max_temperature = T20C + (base_heating * calculated_laser_rating) //573.15K with T1 stock parts
+
+/obj/machinery/atmospherics/components/binary/thermomachine/update_icon_state()
+	if(panel_open)
+		icon_state = icon_state_open
+		return ..()
+	if(on && is_operational)
+		icon_state = icon_state_on
+		return ..()
+	icon_state = icon_state_off
+	return ..()
+
+/obj/machinery/atmospherics/components/binary/thermomachine/update_overlays()
+	. = ..()
+	. += getpipeimage(icon, "pipe", dir, COLOR_LIME, piping_layer)
+	. += getpipeimage(icon, "pipe", turn(dir, 180), COLOR_MOSTLY_PURE_RED, piping_layer)
+	if(holding)
+		var/mutable_appearance/holding = mutable_appearance(icon, "holding")
+		. += holding
+
+/obj/machinery/atmospherics/components/binary/thermomachine/update_icon_nopipes()
+	cut_overlays()
+	if(showpipe)
+		add_overlay(getpipeimage(icon, "scrub_cap", initialize_directions))
```

Comment:
If you're messsing with overlays dont be clearing them yourself like this, put this in update_overlays as they get called around the same time anyway

Reply:
this was the old way to do this, now i moved it to update_overlays()